### PR TITLE
fix(gatsby): use separate eslint-loader for rules that are always required

### LINKED ISSUE #29105: Overwhelmed by ESLint errors during `gatsby develop` after updating to Gatsby 2.31.0
## Description

After updating to the newest version (gatsby@2.31.0) my terminal is flooded by ESLint errors. I've pasted an example below under "Actual result", but the error is repeated for **every** JSX file at once, resulting in the terminal being full of nothing but ESLint errors.

The "undefined rules" mentioned in the terminal are defined in [eslint-config-standard](https://github.com/standard/eslint-config-standard/blob/master/eslintrc.json).

The errors disappear if I force Gatsby to use `eslint@^7` using yarn's resolutions field. This didn't happend before `gatsby@2.31.0`.

### Steps to reproduce

There's a [minimal reproduction here](https://github.com/nikoladev/gatsby-eslint-warnings). Follow the steps in the README (copied below) to see the errors and to see them disappear when using yarn's resolution:

1. Run `yarn` to install all dependencies
2. Run `yarn develop`
    - Look at all the errors in the terminal
3. Stop the develop process
4. In `package.json` change `"resolutions_REMOVETHIS"` to `"resolutions"` and save the changes
5. Run `yarn` to reinstall dependencies
6. Run `yarn develop`
    - There are no errors in the terminal

### Expected result

The develop process should run normally and show no errors.

### Actual result

This is shown in the terminal:

```
 ERROR #98123  WEBPACK

Generating development JavaScript bundle failed


/Users/[...]/gatsby-eslint-warnings/src/pages/index.jsx
  1:1  error  Definition for rule 'default-case-last' was not found         default-case-last
  1:1  error  Definition for rule 'no-useless-backreference' was not found  no-useless-backreference
  1:1  error  Definition for rule 'no-loss-of-precision' was not found      no-loss-of-precision
  1:1  error  Definition for rule 'no-unreachable-loop' was not found       no-unreachable-loop

✖ 4 problems (4 errors, 0 warnings)


File: src/pages/index.jsx


 ERROR #98123  WEBPACK

Generating development JavaScript bundle failed

.eslintrc » eslint-config-standard:
        Environment key "es2021" is unknown


File: .cache/polyfill-entry.js

failed Building development bundle - 3.242s
```

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.

```
  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 14.15.4 - ~/.n/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
  Languages:
    Python: 2.7.17 - /usr/local/bin/python
  Browsers:
    Chrome: 87.0.4280.141
    Firefox: 84.0.2
    Safari: 14.0.2
  npmPackages:
    gatsby: 2.31.0 => 2.31.0 
```

### LINKED ISSUE #29133: gatsby 2.30.0 to 2.31.0 breaks .eslintrc.js
## Description

https://github.com/gatsbyjs/gatsby/commit/5f0d2146b591a0c09576c455f2f4f7e1424f8a20

### Steps to reproduce

Upgrade to gatsby 2.31.1 from 2.31.0 and .eslintrc.js file is ignored

```
  System:
    OS: macOS 11.1
    CPU: (12) x64 Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.15.3 - ~/.nvm/versions/node/v14.15.3/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.11 - ~/.nvm/versions/node/v14.15.3/bin/npm
  Languages:
    Python: 2.7.17 - /usr/local/bin/python
  Browsers:
    Chrome: 88.0.4324.96
    Firefox: 84.0
    Safari: 14.0.2
  npmPackages:
    gatsby: ^2.31.1 => 2.31.1
    gatsby-image: ^2.10.0 => 2.10.0
    gatsby-plugin-canonical-urls: ^2.9.0 => 2.9.0
    gatsby-plugin-disqus: ^1.2.2 => 1.2.2
    gatsby-plugin-google-tagmanager: ^2.10.0 => 2.10.0
    gatsby-plugin-local-search: ^2.0.0 => 2.0.0
    gatsby-plugin-manifest: ^2.11.0 => 2.11.0
    gatsby-plugin-offline: ^3.9.0 => 3.9.0
    gatsby-plugin-postcss: ^3.6.0 => 3.6.0
    gatsby-plugin-react-helmet: ^3.9.0 => 3.9.0
    gatsby-plugin-robots-txt: ^1.5.5 => 1.5.5
    gatsby-plugin-s3: ^0.3.8 => 0.3.8
    gatsby-plugin-sass: ^3.1.0 => 3.1.0
    gatsby-plugin-sharp: ^2.13.1 => 2.13.1
    gatsby-plugin-sitemap: ^2.11.0 => 2.11.0
    gatsby-plugin-svgr: ^2.1.0 => 2.1.0
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.18 => 1.1.18
    gatsby-source-contentful: ^3.1.3 => 3.1.3
    gatsby-source-filesystem: ^2.10.0 => 2.10.0
    gatsby-transformer-remark: ^2.15.0 => 2.15.0
    gatsby-transformer-sharp: ^2.11.0 => 2.11.0
```

### PR DESCRIPTION
**Pull request by vladar**
## Description

Fast-refresh as a hot-reloading mechanism has several constraints. And we need to enforce those constraints using eslint and custom eslint rules. Originally we've implemented this in #28911. And the idea was to merge those rules into a custom user eslint config.

This has caused several issues. This PR suggests a different approach by introducing a separate `eslint-loader` for fast-refresh rules that is completely isolated from the user custom eslint setup.

This loader is only added when there is a custom .eslintrc`. For default Gatsby eslint config we still have a single loader (and just add fast-refresh rules to it).

### Documentation

https://www.gatsbyjs.com/docs/how-to/custom-configuration/eslint/

(this PR warrants updating those docs a bit)

## Related Issues

An alternative approach to #28911

Fixes #29105
Fixes #29133

### CODE REVIEW

**Comment by vladar** [comment]
Closing this in favor of a fix with a smaller change surface: #29319

File: packages/gatsby/src/utils/webpack.config.js [hunk 1]
```diff
@@ -339,11 +339,21 @@ module.exports = async (
...
           configRules = configRules.concat([rules.eslint(schema)])
         }
 
+        // Enforce fast-refresh rules even with local eslint config
+        if (
+          isCustomEslint &&
+          process.env.GATSBY_HOT_LOADER === `fast-refresh`
+        ) {
+          configRules = configRules.concat([rules.eslintRequired()])
+        }
```
Comment (vladar) [inline packages/gatsby/src/utils/webpack.config.js:355]: Basically the main piece. We add another minimal loader only when there is user-defined eslint config. When there is no custom config - we use a single default loader (with fast-refresh rules added to it)

**Review by pieh — APPROVED** [review]
I tried testing as much combinations from reports as possible - with `gatsby-plugin-eslint` - both v2 (that use `eslint-loader`) and v3 (that use `eslint-webpack-plugin`) and without

using optional chaining (for old parser), using `eslint-disable` with "not existing rules" and from what I can tell I had expected results, so let's merge this - at very least this will fix known issues.

Huge thanks!

**Comment by vladar** [comment]
Published in `gatsby@2.32.1`

**Pull request merged**